Feature/category

FAEZ10 opened the pull request:
# Pull Request: **Category Feature**

## Type of Change

- [x] Feature
- [ ] Fix

## Description

This PR introduces the **category feature**, allowing users to classify and organize items effectively. The feature addresses the following use case: [link or description of the related issue].  

### Summary of Changes:
- Created a `Category` model with attributes (e.g., `id`, `name`, `description`).
- Added endpoints for category CRUD operations (`create`, `read`, `update`, `delete`).
- Integrated category assignment within the existing workflow.
- Updated the database schema to support the new category structure.
- Implemented UI components to display and manage categories.

### Dependencies:
This feature requires the following dependencies:
- Updated migration files for the database schema.
- Any other specific library or tool dependencies (if applicable).

## How Has This Been Tested?

- [x] **Unit Tests**: Tested category creation, updating, and deletion logic.
- [x] **Integration Tests**: Verified category integration within the workflow.
- [ ] **Manual Testing**: Performed end-to-end testing to ensure UI consistency.

### Test Configuration:
- Database: [specific version]
- Framework: [specific version]
- Relevant tools: [e.g., Postman for API testing]

## Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect these changes.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my feature works as expected.
- [x] All new and existing tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

## Screenshots or Demo (if applicable):
*(Include any relevant screenshots, GIFs, or links to the feature in action)*

DaniOmer commented on src/apps/product/api/category/category.route.ts:
```diff
@@ -0,0 +1,181 @@
+router.post(
+  "/",
+  validateDtoMiddleware(CategoryCreateDTO),
+  categoryController.create.bind(categoryController)
```
Il manque le middleware pour restreindre la route à l'admin uniquement

FAEZ10 replied:
Ça marche je vais faire ça je vais profiter pour le faire pour le reste des PR

DaniOmer replied:
super merci

DaniOmer commented on src/apps/product/api/category/category.route.ts:
```diff
@@ -0,0 +1,181 @@
+  "/:id",
+  validateIdMiddleware("Category"),
+  validateDtoMiddleware(CategoryUpdateDTO),
+  categoryController.updateById.bind(categoryController)
```
Il manque le middleware pour restreindre la route à l'admin uniquement

DaniOmer commented on src/apps/product/api/category/category.route.ts:
```diff
@@ -0,0 +1,181 @@
+router.delete(
+  "/:id",
+  validateIdMiddleware("Category"),
+  categoryController.deleteById.bind(categoryController)
```
Il manque le middleware pour restreindre la route à l'admin uniquement

DaniOmer commented on src/apps/product/domain/category/category.service.ts:
```diff
@@ -0,0 +1,54 @@
+  async getCategoryById(id: string): Promise<ICategory> {
+    const category = await this.repository.findByIdWithParent(id);
+    return this.validateDataExists(category, id);
+  }
```
par endroits tu utilise `this.validateDataExists` pour throw l'erreur et en même temps `throw new BadRequestError` il faut vraiment uniformiser ce truc sinon mieux vaut le retirer. Avec cette implémentation ça laisse croire que ça fait autre chose que de throw une erreur

FAEZ10 replied:
Boss c'est du détail et je pense j'ai fait la même implémentation partout

DaniOmer replied:
Ouep je sais que c'est du détails quand on code, mais ça peut faire la diff quand tu debug. surtout si c'est un autre dev. mais sinon comme tu veux

DaniOmer commented on src/apps/product/domain/category/category.service.ts:
```diff
@@ -0,0 +1,54 @@
+      }
+    }
+    const updatedCategory = await this.repository.updateById(id, updates);
+    return this.validateDataExists(updatedCategory, id);
```
pareil ici

DaniOmer commented on src/index.ts:
```diff
@@ -39,6 +40,10 @@ async function startApp() {
 
+  
+    // category routes
+    app.use("/api/categories", categoryRoute);
```
productRouter serait mieux pour rester dans la même nomenclature que les autres

FAEZ10 replied:
Oui je sais je me suis rendu compte après t'inquiète je vais le modifier

DaniOmer commented:
Good job dans l'ensemble 👍

DaniOmer commented:
Good job dans l'ensemble 👍

DaniOmer commented on src/apps/product/domain/category/category.dto.ts:
```diff
@@ -0,0 +1,33 @@
+
+  @IsOptional()
+  @IsString()
+  readonly parentId?: string;
```
Tu peux utiliser IsMongoId ici pour la validation

DaniOmer commented on src/apps/product/domain/category/category.service.ts:
```diff
@@ -0,0 +1,54 @@
+      if (!parentCategory) {
+        throw new BadRequestError({
+          message: `Parent category not found for ID: ${data.parentId}`,
+          code: 404,
```
ce serait pas mal de rajouter un context ici : genre create_category : "Invalid parent id"

FAEZ14 committed:
feat: add auth and role-check middlewares with improved error handling for Categories module

- Added `authenticateMiddleware` to secure category routes.
- Integrated `checkRoleMiddleware` to restrict access based on user roles (e.g., admin).
- Enhanced error handling in `CategoryService` by including contextual information in `BadRequestError` (e.g., invalid parent category ID).

FAEZ14 committed:
feat: add IsMongoId validation to Category DTOs

- Added `@IsMongoId()` decorator to `parentId` in `CategoryCreateDTO` and `CategoryUpdateDTO`.
- Ensures that `parentId` is a valid MongoDB ObjectId when provided.
- Improves data validation and prevents invalid IDs from being processed in the Categories module.

DaniOmer commented on src/apps/product/domain/category/category.service.ts:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  async createCategory(data: Partial<ICategory>): Promise<ICategory> {
+    if (data.parentId) {
```
Hello @FAEZ10 ici au lieu de `Partial<ICategory>` il serait mieux d'utiliser `CategoryCreateDTO` pour une validation strict et cohérente. Il y a un exemple dans `PermissionService` merci

FAEZ10 committed:
Merge branch 'dev' into feature/category
